snapcraft/internal/project_loader/_extensions/_desktop_common.py

```diff
@@ -0,0 +1,80 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from typing import Any, Dict
+
+from ._extension import Extension
+
+
+class DesktopCommonExtension(Extension):
+    """The Common Desktop extension.
+    This extension is to be subclassed by the gnome and kde extensions.
+    It includes common code that each desktop extension would otherwise
+    need to duplicate.
```

Comment:
Nice abstraction.

Also, the whole file here feels quite clean and readable. It's not clear how the underlying Extension type will handle that data (will it override or append? when? etc), but it looks good indeed.

Reply:
The idea now is to just append. The base class settings will be accessible in the subclasses allowing them to either replace or augment as they desire. The way I have done this in the gnome extension is to append the gnome-specific values via code like:

```python
        self.parts = {
            **self.parts,
            "gnome-extension": {
                "after": after_dependencies,
                "plugin": "dump",
                "source": "$SNAPCRAFT_EXTENSIONS_DIR/gnome",
                "source-type": "local",
                "organize": {"desktop-*": "snap/command-chain/"},
                "build-packages": ["build-essential", "libgtk-3-dev"],
            },
        }
```

Here you see I directly reference the `self.parts` when constructing a replacement value to insert the previous content as-is into the new value. This results in the equivalent combined definition of:

```python
        self.parts = {
            "desktop-common-extension": { # expanded from **self.parts when setting new value
                "plugin": "dump",
                "source": "$SNAPCRAFT_EXTENSIONS_DIR/desktop-common",
                "organize": {"desktop-*": "snap/command-chain/"},
            },
            "desktop-common-bindtextdomain": { # expanded from **self.parts when setting new value
                "plugin": "make",
                "source": "$SNAPCRAFT_EXTENSIONS_DIR/bindtextdomain",
            },
            "gnome-extension": { # defined explicitly when setting new value
                "after": after_dependencies,
                "plugin": "dump",
                "source": "$SNAPCRAFT_EXTENSIONS_DIR/gnome",
                "source-type": "local",
                "organize": {"desktop-*": "snap/command-chain/"},
                "build-packages": ["build-essential", "libgtk-3-dev"],
            },
        }
```